Sync shows JobCancellationException to user

Maybe / probably because of the latest changes to make `SyncManager` more compatible with coroutines

~~~
EXCEPTION
kotlinx.coroutines.JobCancellationException: Parent job is Cancelling; job=NonCancellable
Caused by: java.lang.InterruptedException
	at kotlinx.coroutines.JobKt.runBlocking(Unknown Source:132)
	at kotlinx.coroutines.JobKt.runBlocking$default(Unknown Source:3)
	at at.bitfire.davdroid.sync.JtxSyncer.syncCollection(SourceFile:10)
	at at.bitfire.davdroid.sync.JtxSyncer.syncCollection(SourceFile:1)
	at at.bitfire.davdroid.sync.Syncer.syncCollectionContents$davx5_404110000_4_4_11_alpha_1_gplayRelease(SourceFile:44)
	at at.bitfire.davdroid.sync.Syncer.sync$davx5_404110000_4_4_11_alpha_1_gplayRelease(SourceFile:42)
	at at.bitfire.davdroid.sync.Syncer.invoke(SourceFile:211)
	at at.bitfire.davdroid.sync.worker.BaseSyncWorker.doSyncWork$lambda$0(SourceFile:1)
	at at.bitfire.davdroid.sync.worker.BaseSyncWorker.$r8$lambda$PdhGR9yaeBRQyLwc6Ku-xpX_yQY(SourceFile:1)
	at AccountScreenKt$$ExternalSyntheticLambda14.invoke(SourceFile:134)
	at kotlinx.coroutines.InterruptibleKt$runInterruptible$2.invokeSuspend(SourceFile:59)
	at kotlinx.coroutines.InterruptibleKt$runInterruptible$2.invoke(SourceFile:13)
	at androidx.tracing.Trace.startUndspatched(SourceFile:17)
	at kotlinx.coroutines.JobKt.withContext(Unknown Source:46)
	at kotlinx.coroutines.JobKt.runInterruptible(Unknown Source:7)
	at at.bitfire.davdroid.sync.worker.BaseSyncWorker.doSyncWork(SourceFile:345)
	at at.bitfire.davdroid.sync.worker.BaseSyncWorker.doWork$suspendImpl(SourceFile:451)
	at at.bitfire.davdroid.sync.worker.BaseSyncWorker.doWork(Unknown Source:0)
	at androidx.work.CoroutineWorker$startWork$1.invokeSuspend(SourceFile:29)
	at androidx.work.CoroutineWorker$startWork$1.invoke(SourceFile:13)
	at androidx.work.ListenableFutureKt$launchFuture$1$2.invokeSuspend(SourceFile:37)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(SourceFile:9)
	at kotlinx.coroutines.DispatchedTask.run(SourceFile:115)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(SourceFile:127)
~~~

During network switch

--- a/app/src/androidTest/kotlin/at/bitfire/davdroid/sync/TestSyncManager.kt
+++ b/app/src/androidTest/kotlin/at/bitfire/davdroid/sync/TestSyncManager.kt
@@ -63,7 +63,7 @@ class TestSyncManager @AssistedInject constructor(
     }
 
     var didQueryCapabilities = false
-    override fun queryCapabilities(): SyncState? {
+    override suspend fun queryCapabilities(): SyncState? {
         if (didQueryCapabilities)
             throw IllegalStateException("queryCapabilities() must not be called twice")
         didQueryCapabilities = true
@@ -89,7 +89,7 @@ class TestSyncManager @AssistedInject constructor(
 
     var listAllRemoteResult = emptyList<Pair<Response, Response.HrefRelation>>()
     var didListAllRemote = false
-    override fun listAllRemote(callback: MultiResponseCallback) {
+    override suspend fun listAllRemote(callback: MultiResponseCallback) {
         if (didListAllRemote)
             throw IllegalStateException("listAllRemote() must not be called twice")
         didListAllRemote = true
@@ -99,7 +99,7 @@ class TestSyncManager @AssistedInject constructor(
 
     var assertDownloadRemote = emptyMap<HttpUrl, String>()
     var didDownloadRemote = false
-    override fun downloadRemote(bunch: List<HttpUrl>) {
+    override suspend fun downloadRemote(bunch: List<HttpUrl>) {
         didDownloadRemote = true
         assertEquals(assertDownloadRemote.keys.toList(), bunch)
 

--- a/app/src/main/kotlin/at/bitfire/davdroid/sync/CalendarSyncManager.kt
+++ b/app/src/main/kotlin/at/bitfire/davdroid/sync/CalendarSyncManager.kt
@@ -34,6 +34,7 @@ import dagger.assisted.Assisted
 import dagger.assisted.AssistedFactory
 import dagger.assisted.AssistedInject
 import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.runInterruptible
 import net.fortuna.ical4j.model.Component
 import net.fortuna.ical4j.model.component.VAlarm
 import net.fortuna.ical4j.model.property.Action
@@ -99,19 +100,21 @@ class CalendarSyncManager @AssistedInject constructor(
         return true
     }
 
-    override fun queryCapabilities(): SyncState? =
-        SyncException.wrapWithRemoteResource(collection.url) {
+    override suspend fun queryCapabilities(): SyncState? =
+        SyncException.wrapWithRemoteResourceSuspending(collection.url) {
             var syncState: SyncState? = null
-            davCollection.propfind(0, MaxResourceSize.NAME, SupportedReportSet.NAME, GetCTag.NAME, SyncToken.NAME) { response, relation ->
-                if (relation == Response.HrefRelation.SELF) {
-                    response[MaxResourceSize::class.java]?.maxSize?.let { maxSize ->
-                        logger.info("Calendar accepts events up to ${Formatter.formatFileSize(context, maxSize)}")
-                    }
+            runInterruptible {
+                davCollection.propfind(0, MaxResourceSize.NAME, SupportedReportSet.NAME, GetCTag.NAME, SyncToken.NAME) { response, relation ->
+                    if (relation == Response.HrefRelation.SELF) {
+                        response[MaxResourceSize::class.java]?.maxSize?.let { maxSize ->
+                            logger.info("Calendar accepts events up to ${Formatter.formatFileSize(context, maxSize)}")
+                        }
 
-                    response[SupportedReportSet::class.java]?.let { supported ->
-                        hasCollectionSync = supported.reports.contains(SupportedReportSet.SYNC_COLLECTION)
+                        response[SupportedReportSet::class.java]?.let { supported ->
+                            hasCollectionSync = supported.reports.contains(SupportedReportSet.SYNC_COLLECTION)
+                        }
+                        syncState = syncState(response)
                     }
-                    syncState = syncState(response)
                 }
             }
 
@@ -184,15 +187,17 @@ class CalendarSyncManager @AssistedInject constructor(
             os.toByteArray().toRequestBody(DavCalendar.MIME_ICALENDAR_UTF8)
         }
 
-    override fun listAllRemote(callback: MultiResponseCallback) {
+    override suspend fun listAllRemote(callback: MultiResponseCallback) {
         // calculate time range limits
         val limitStart = accountSettings.getTimeRangePastDays()?.let { pastDays ->
             ZonedDateTime.now().minusDays(pastDays.toLong()).toInstant()
         }
 
-        return SyncException.wrapWithRemoteResource(collection.url) {
+        return SyncException.wrapWithRemoteResourceSuspending(collection.url) {
             logger.info("Querying events since $limitStart")
-            davCollection.calendarQuery(Component.VEVENT, limitStart, null, callback)
+            runInterruptible {
+                davCollection.calendarQuery(Component.VEVENT, limitStart, null, callback)
+            }
         }
     }
 

--- a/app/src/main/kotlin/at/bitfire/davdroid/sync/CalendarSyncManager.kt
+++ b/app/src/main/kotlin/at/bitfire/davdroid/sync/CalendarSyncManager.kt
@@ -201,45 +201,47 @@ class CalendarSyncManager @AssistedInject constructor(
         }
     }
 
-    override fun downloadRemote(bunch: List<HttpUrl>) {
+    override suspend fun downloadRemote(bunch: List<HttpUrl>) {
         logger.info("Downloading ${bunch.size} iCalendars: $bunch")
-        SyncException.wrapWithRemoteResource(collection.url) {
-            davCollection.multiget(bunch) { response, _ ->
-                /*
-                 * Real-world servers may return:
-                 *
-                 * - unrelated resources
-                 * - the collection itself
-                 * - the requested resources, but with a different collection URL (for instance, `/cal/1.ics` instead of `/shared-cal/1.ics`).
-                 *
-                 * So we:
-                 *
-                 * - ignore unsuccessful responses,
-                 * - ignore responses without requested calendar data (should also ignore collections and hopefully unrelated resources), and
-                 * - take the last segment of the href as the file name and assume that it's in the requested collection.
-                 */
-                SyncException.wrapWithRemoteResource(response.href) wrapResource@ {
-                    if (!response.isSuccess()) {
-                        logger.warning("Ignoring non-successful multi-get response for ${response.href}")
-                        return@wrapResource
+        SyncException.wrapWithRemoteResourceSuspending(collection.url) {
+            runInterruptible {
+                davCollection.multiget(bunch) { response, _ ->
+                    /*
+                     * Real-world servers may return:
+                     *
+                     * - unrelated resources
+                     * - the collection itself
+                     * - the requested resources, but with a different collection URL (for instance, `/cal/1.ics` instead of `/shared-cal/1.ics`).
+                     *
+                     * So we:
+                     *
+                     * - ignore unsuccessful responses,
+                     * - ignore responses without requested calendar data (should also ignore collections and hopefully unrelated resources), and
+                     * - take the last segment of the href as the file name and assume that it's in the requested collection.
+                     */
+                    SyncException.wrapWithRemoteResource(response.href) wrapResource@{
+                        if (!response.isSuccess()) {
+                            logger.warning("Ignoring non-successful multi-get response for ${response.href}")
+                            return@wrapResource
+                        }
+
+                        val iCal = response[CalendarData::class.java]?.iCalendar
+                        if (iCal == null) {
+                            logger.warning("Ignoring multi-get response without calendar-data")
+                            return@wrapResource
+                        }
+
+                        val eTag = response[GetETag::class.java]?.eTag
+                            ?: throw DavException("Received multi-get response without ETag")
+                        val scheduleTag = response[ScheduleTag::class.java]?.scheduleTag
+
+                        processVEvent(
+                            response.href.lastSegment,
+                            eTag,
+                            scheduleTag,
+                            StringReader(iCal)
+                        )
                     }
-
-                    val iCal = response[CalendarData::class.java]?.iCalendar
-                    if (iCal == null) {
-                        logger.warning("Ignoring multi-get response without calendar-data")
-                        return@wrapResource
-                    }
-
-                    val eTag = response[GetETag::class.java]?.eTag
-                        ?: throw DavException("Received multi-get response without ETag")
-                    val scheduleTag = response[ScheduleTag::class.java]?.scheduleTag
-
-                    processVEvent(
-                        response.href.lastSegment,
-                        eTag,
-                        scheduleTag,
-                        StringReader(iCal)
-                    )
                 }
             }
         }

--- a/app/src/main/kotlin/at/bitfire/davdroid/sync/CalendarSyncManager.kt
+++ b/app/src/main/kotlin/at/bitfire/davdroid/sync/CalendarSyncManager.kt
@@ -125,7 +125,7 @@ class CalendarSyncManager @AssistedInject constructor(
         else
             SyncAlgorithm.COLLECTION_SYNC
 
-    override fun processLocallyDeleted(): Boolean {
+    override suspend fun processLocallyDeleted(): Boolean {
         if (localCollection.readOnly) {
             var modified = false
             for (event in localCollection.findDeleted()) {

--- a/app/src/main/kotlin/at/bitfire/davdroid/sync/ContactsSyncManager.kt
+++ b/app/src/main/kotlin/at/bitfire/davdroid/sync/ContactsSyncManager.kt
@@ -202,7 +202,7 @@ class ContactsSyncManager @AssistedInject constructor(
         else
             SyncAlgorithm.PROPFIND_REPORT
 
-    override fun processLocallyDeleted() =
+    override suspend fun processLocallyDeleted() =
             if (localCollection.readOnly) {
                 var modified = false
                 for (group in localCollection.findDeletedGroups()) {

--- a/app/src/main/kotlin/at/bitfire/davdroid/sync/ContactsSyncManager.kt
+++ b/app/src/main/kotlin/at/bitfire/davdroid/sync/ContactsSyncManager.kt
@@ -307,9 +307,9 @@ class ContactsSyncManager @AssistedInject constructor(
             }
         }
 
-    override fun downloadRemote(bunch: List<HttpUrl>) {
+    override suspend fun downloadRemote(bunch: List<HttpUrl>) {
         logger.info("Downloading ${bunch.size} vCard(s): $bunch")
-        SyncException.wrapWithRemoteResource(collection.url) {
+        SyncException.wrapWithRemoteResourceSuspending(collection.url) {
             val contentType: String?
             val version: String?
             when {
@@ -326,35 +326,37 @@ class ContactsSyncManager @AssistedInject constructor(
                     version = null     // 3.0 is the default version; don't request 3.0 explicitly because maybe some vCard3-only servers don't understand it
                 }
             }
-            davCollection.multiget(bunch, contentType, version) { response, _ ->
-                // See CalendarSyncManager for more information about the multi-get response
-                SyncException.wrapWithRemoteResource(response.href) wrapResource@ {
-                    if (!response.isSuccess()) {
-                        logger.warning("Ignoring non-successful multi-get response for ${response.href}")
-                        return@wrapResource
+            runInterruptible {
+                davCollection.multiget(bunch, contentType, version) { response, _ ->
+                    // See CalendarSyncManager for more information about the multi-get response
+                    SyncException.wrapWithRemoteResource(response.href) wrapResource@{
+                        if (!response.isSuccess()) {
+                            logger.warning("Ignoring non-successful multi-get response for ${response.href}")
+                            return@wrapResource
+                        }
+
+                        val card = response[AddressData::class.java]?.card
+                        if (card == null) {
+                            logger.warning("Ignoring multi-get response without address-data")
+                            return@wrapResource
+                        }
+
+                        val eTag = response[GetETag::class.java]?.eTag
+                            ?: throw DavException("Received multi-get response without ETag")
+
+                        var isJCard = hasJCard      // assume that server has sent what we have requested (we ask for jCard only when the server advertises it)
+                        response[GetContentType::class.java]?.type?.let { type ->
+                            isJCard = type.sameTypeAs(DavUtils.MEDIA_TYPE_JCARD)
+                        }
+
+                        processCard(
+                            response.href.lastSegment,
+                            eTag,
+                            StringReader(card),
+                            isJCard,
+                            resourceDownloader
+                        )
                     }
-
-                    val card = response[AddressData::class.java]?.card
-                    if (card == null) {
-                        logger.warning("Ignoring multi-get response without address-data")
-                        return@wrapResource
-                    }
-
-                    val eTag = response[GetETag::class.java]?.eTag
-                        ?: throw DavException("Received multi-get response without ETag")
-
-                    var isJCard = hasJCard      // assume that server has sent what we have requested (we ask for jCard only when the server advertises it)
-                    response[GetContentType::class.java]?.type?.let { type ->
-                        isJCard = type.sameTypeAs(DavUtils.MEDIA_TYPE_JCARD)
-                    }
-
-                    processCard(
-                        response.href.lastSegment,
-                        eTag,
-                        StringReader(card),
-                        isJCard,
-                        resourceDownloader
-                    )
                 }
             }
         }

--- a/app/src/main/kotlin/at/bitfire/davdroid/sync/ContactsSyncManager.kt
+++ b/app/src/main/kotlin/at/bitfire/davdroid/sync/ContactsSyncManager.kt
@@ -46,6 +46,7 @@ import dagger.assisted.AssistedInject
 import ezvcard.VCardVersion
 import ezvcard.io.CannotParseException
 import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.runInterruptible
 import okhttp3.HttpUrl
 import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
 import okhttp3.MediaType
@@ -166,25 +167,27 @@ class ContactsSyncManager @AssistedInject constructor(
         return true
     }
 
-    override fun queryCapabilities(): SyncState? {
-        return SyncException.wrapWithRemoteResource(collection.url) {
+    override suspend fun queryCapabilities(): SyncState? {
+        return SyncException.wrapWithRemoteResourceSuspending(collection.url) {
             var syncState: SyncState? = null
-            davCollection.propfind(0, MaxResourceSize.NAME, SupportedAddressData.NAME, SupportedReportSet.NAME, GetCTag.NAME, SyncToken.NAME) { response, relation ->
-                if (relation == Response.HrefRelation.SELF) {
-                    response[MaxResourceSize::class.java]?.maxSize?.let { maxSize ->
-                        logger.info("Address book accepts vCards up to ${Formatter.formatFileSize(context, maxSize)}")
-                    }
+            runInterruptible {
+                davCollection.propfind(0, MaxResourceSize.NAME, SupportedAddressData.NAME, SupportedReportSet.NAME, GetCTag.NAME, SyncToken.NAME) { response, relation ->
+                    if (relation == Response.HrefRelation.SELF) {
+                        response[MaxResourceSize::class.java]?.maxSize?.let { maxSize ->
+                            logger.info("Address book accepts vCards up to ${Formatter.formatFileSize(context, maxSize)}")
+                        }
 
-                    response[SupportedAddressData::class.java]?.let { supported ->
-                        hasVCard4 = supported.hasVCard4()
+                        response[SupportedAddressData::class.java]?.let { supported ->
+                            hasVCard4 = supported.hasVCard4()
 
-                        // temporarily disable jCard because of https://github.com/nextcloud/server/issues/29693
-                        // hasJCard = supported.hasJCard()
+                            // temporarily disable jCard because of https://github.com/nextcloud/server/issues/29693
+                            // hasJCard = supported.hasJCard()
+                        }
+                        response[SupportedReportSet::class.java]?.let { supported ->
+                            hasCollectionSync = supported.reports.contains(SupportedReportSet.SYNC_COLLECTION)
+                        }
+                        syncState = syncState(response)
                     }
-                    response[SupportedReportSet::class.java]?.let { supported ->
-                        hasCollectionSync = supported.reports.contains(SupportedReportSet.SYNC_COLLECTION)
-                    }
-                    syncState = syncState(response)
                 }
             }
 
@@ -297,9 +300,11 @@ class ContactsSyncManager @AssistedInject constructor(
             return@wrapWithLocalResource os.toByteArray().toRequestBody(mimeType)
         }
 
-    override fun listAllRemote(callback: MultiResponseCallback) =
-        SyncException.wrapWithRemoteResource(collection.url) {
-            davCollection.propfind(1, ResourceType.NAME, GetETag.NAME, callback = callback)
+    override suspend fun listAllRemote(callback: MultiResponseCallback) =
+        SyncException.wrapWithRemoteResourceSuspending(collection.url) {
+            runInterruptible {
+                davCollection.propfind(1, ResourceType.NAME, GetETag.NAME, callback = callback)
+            }
         }
 
     override fun downloadRemote(bunch: List<HttpUrl>) {

--- a/app/src/main/kotlin/at/bitfire/davdroid/sync/JtxSyncManager.kt
+++ b/app/src/main/kotlin/at/bitfire/davdroid/sync/JtxSyncManager.kt
@@ -125,28 +125,30 @@ class JtxSyncManager @AssistedInject constructor(
         }
     }
 
-    override fun downloadRemote(bunch: List<HttpUrl>) {
+    override suspend fun downloadRemote(bunch: List<HttpUrl>) {
         logger.info("Downloading ${bunch.size} iCalendars: $bunch")
         // multiple iCalendars, use calendar-multi-get
-        SyncException.wrapWithRemoteResource(collection.url) {
-            davCollection.multiget(bunch) { response, _ ->
-                // See CalendarSyncManager for more information about the multi-get response
-                SyncException.wrapWithRemoteResource(response.href) wrapResource@ {
-                    if (!response.isSuccess()) {
-                        logger.warning("Ignoring non-successful multi-get response for ${response.href}")
-                        return@wrapResource
+        SyncException.wrapWithRemoteResourceSuspending(collection.url) {
+            runInterruptible {
+                davCollection.multiget(bunch) { response, _ ->
+                    // See CalendarSyncManager for more information about the multi-get response
+                    SyncException.wrapWithRemoteResource(response.href) wrapResource@{
+                        if (!response.isSuccess()) {
+                            logger.warning("Ignoring non-successful multi-get response for ${response.href}")
+                            return@wrapResource
+                        }
+
+                        val iCal = response[CalendarData::class.java]?.iCalendar
+                        if (iCal == null) {
+                            logger.warning("Ignoring multi-get response without calendar-data")
+                            return@wrapResource
+                        }
+
+                        val eTag = response[GetETag::class.java]?.eTag
+                            ?: throw DavException("Received multi-get response without ETag")
+
+                        processICalObject(response.href.lastSegment, eTag, StringReader(iCal))
                     }
-
-                    val iCal = response[CalendarData::class.java]?.iCalendar
-                    if (iCal == null) {
-                        logger.warning("Ignoring multi-get response without calendar-data")
-                        return@wrapResource
-                    }
-
-                    val eTag = response[GetETag::class.java]?.eTag
-                        ?: throw DavException("Received multi-get response without ETag")
-
-                    processICalObject(response.href.lastSegment, eTag, StringReader(iCal))
                 }
             }
         }

--- a/app/src/main/kotlin/at/bitfire/davdroid/sync/JtxSyncManager.kt
+++ b/app/src/main/kotlin/at/bitfire/davdroid/sync/JtxSyncManager.kt
@@ -31,6 +31,7 @@ import dagger.assisted.Assisted
 import dagger.assisted.AssistedFactory
 import dagger.assisted.AssistedInject
 import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.runInterruptible
 import okhttp3.HttpUrl
 import okhttp3.RequestBody
 import okhttp3.RequestBody.Companion.toRequestBody
@@ -79,16 +80,18 @@ class JtxSyncManager @AssistedInject constructor(
         return true
     }
 
-    override fun queryCapabilities() =
-        SyncException.wrapWithRemoteResource(collection.url) {
+    override suspend fun queryCapabilities() =
+        SyncException.wrapWithRemoteResourceSuspending(collection.url) {
             var syncState: SyncState? = null
-            davCollection.propfind(0, GetCTag.NAME, MaxResourceSize.NAME, SyncToken.NAME) { response, relation ->
-                if (relation == Response.HrefRelation.SELF) {
-                    response[MaxResourceSize::class.java]?.maxSize?.let { maxSize ->
-                        logger.info("Collection accepts resources up to ${Formatter.formatFileSize(context, maxSize)}")
-                    }
+            runInterruptible {
+                davCollection.propfind(0, GetCTag.NAME, MaxResourceSize.NAME, SyncToken.NAME) { response, relation ->
+                    if (relation == Response.HrefRelation.SELF) {
+                        response[MaxResourceSize::class.java]?.maxSize?.let { maxSize ->
+                            logger.info("Collection accepts resources up to ${Formatter.formatFileSize(context, maxSize)}")
+                        }
 
-                    syncState = syncState(response)
+                        syncState = syncState(response)
+                    }
                 }
             }
             syncState
@@ -104,16 +107,20 @@ class JtxSyncManager @AssistedInject constructor(
 
     override fun syncAlgorithm() = SyncAlgorithm.PROPFIND_REPORT
 
-    override fun listAllRemote(callback: MultiResponseCallback) {
-        SyncException.wrapWithRemoteResource(collection.url) {
+    override suspend fun listAllRemote(callback: MultiResponseCallback) {
+        SyncException.wrapWithRemoteResourceSuspending(collection.url) {
             if (localCollection.supportsVTODO) {
                 logger.info("Querying tasks")
-                davCollection.calendarQuery("VTODO", null, null, callback)
+                runInterruptible {
+                    davCollection.calendarQuery("VTODO", null, null, callback)
+                }
             }
 
             if (localCollection.supportsVJOURNAL) {
                 logger.info("Querying journals")
-                davCollection.calendarQuery("VJOURNAL", null, null, callback)
+                runInterruptible {
+                    davCollection.calendarQuery("VJOURNAL", null, null, callback)
+                }
             }
         }
     }

--- a/app/src/main/kotlin/at/bitfire/davdroid/sync/SyncException.kt
+++ b/app/src/main/kotlin/at/bitfire/davdroid/sync/SyncException.kt
@@ -26,6 +26,22 @@ class SyncException(cause: Throwable): Exception(cause) {
                 throw e
             } catch (e: Throwable) {
                 throw
+                if (localResource != null)
+                    SyncException(e).setLocalResourceIfNull(localResource)
+                else
+                    e
+            }
+        }
+
+        suspend fun<T> wrapWithLocalResourceSuspending(localResource: LocalResource<*>?, body: suspend () -> T): T {
+            try {
+                return body()
+            } catch (e: SyncException) {
+                if (localResource != null)
+                    e.setLocalResourceIfNull(localResource)
+                throw e
+            } catch (e: Throwable) {
+                throw
                     if (localResource != null)
                         SyncException(e).setLocalResourceIfNull(localResource)
                     else
@@ -46,6 +62,22 @@ class SyncException(cause: Throwable): Exception(cause) {
                         SyncException(e).setRemoteResourceIfNull(remoteResource)
                     else
                         e
+            }
+        }
+
+        suspend fun<T> wrapWithRemoteResourceSuspending(remoteResource: HttpUrl?, body: suspend () -> T): T {
+            try {
+                return body()
+            } catch (e: SyncException) {
+                if (remoteResource != null)
+                    e.setRemoteResourceIfNull(remoteResource)
+                throw e
+            } catch (e: Throwable) {
+                throw
+                if (remoteResource != null)
+                    SyncException(e).setRemoteResourceIfNull(remoteResource)
+                else
+                    e
             }
         }
 

--- a/app/src/main/kotlin/at/bitfire/davdroid/sync/SyncException.kt
+++ b/app/src/main/kotlin/at/bitfire/davdroid/sync/SyncException.kt
@@ -11,13 +11,13 @@ import okhttp3.HttpUrl
  * Exception that wraps another notification together with potential information about
  * a local and/or remote resource that is related to the exception.
  */
-class SyncException(cause: Throwable): Exception(cause) {
+class SyncException(cause: Throwable) : Exception(cause) {
 
     companion object {
 
         // provide lambda wrappers for setting the local/remote resource
 
-        fun<T> wrapWithLocalResource(localResource: LocalResource<*>?, body: () -> T): T {
+        fun <T> wrapWithLocalResource(localResource: LocalResource<*>?, body: () -> T): T {
             try {
                 return body()
             } catch (e: SyncException) {
@@ -25,15 +25,14 @@ class SyncException(cause: Throwable): Exception(cause) {
                     e.setLocalResourceIfNull(localResource)
                 throw e
             } catch (e: Throwable) {
-                throw
-                if (localResource != null)
+                throw if (localResource != null)
                     SyncException(e).setLocalResourceIfNull(localResource)
                 else
                     e
             }
         }
 
-        suspend fun<T> wrapWithLocalResourceSuspending(localResource: LocalResource<*>?, body: suspend () -> T): T {
+        suspend fun <T> wrapWithLocalResourceSuspending(localResource: LocalResource<*>?, body: suspend () -> T): T {
             try {
                 return body()
             } catch (e: SyncException) {
@@ -41,15 +40,14 @@ class SyncException(cause: Throwable): Exception(cause) {
                     e.setLocalResourceIfNull(localResource)
                 throw e
             } catch (e: Throwable) {
-                throw
-                    if (localResource != null)
-                        SyncException(e).setLocalResourceIfNull(localResource)
-                    else
-                        e
+                throw if (localResource != null)
+                    SyncException(e).setLocalResourceIfNull(localResource)
+                else
+                    e
             }
         }
 
-        fun<T> wrapWithRemoteResource(remoteResource: HttpUrl?, body: () -> T): T {
+        fun <T> wrapWithRemoteResource(remoteResource: HttpUrl?, body: () -> T): T {
             try {
                 return body()
             } catch (e: SyncException) {
@@ -57,15 +55,14 @@ class SyncException(cause: Throwable): Exception(cause) {
                     e.setRemoteResourceIfNull(remoteResource)
                 throw e
             } catch (e: Throwable) {
-                throw
-                    if (remoteResource != null)
-                        SyncException(e).setRemoteResourceIfNull(remoteResource)
-                    else
-                        e
+                throw if (remoteResource != null)
+                    SyncException(e).setRemoteResourceIfNull(remoteResource)
+                else
+                    e
             }
         }
 
-        suspend fun<T> wrapWithRemoteResourceSuspending(remoteResource: HttpUrl?, body: suspend () -> T): T {
+        suspend fun <T> wrapWithRemoteResourceSuspending(remoteResource: HttpUrl?, body: suspend () -> T): T {
             try {
                 return body()
             } catch (e: SyncException) {
@@ -73,8 +70,7 @@ class SyncException(cause: Throwable): Exception(cause) {
                     e.setRemoteResourceIfNull(remoteResource)
                 throw e
             } catch (e: Throwable) {
-                throw
-                if (remoteResource != null)
+                throw if (remoteResource != null)
                     SyncException(e).setRemoteResourceIfNull(remoteResource)
                 else
                     e

--- a/app/src/main/kotlin/at/bitfire/davdroid/sync/SyncException.kt
+++ b/app/src/main/kotlin/at/bitfire/davdroid/sync/SyncException.kt
@@ -5,6 +5,7 @@
 package at.bitfire.davdroid.sync
 
 import at.bitfire.davdroid.resource.LocalResource
+import kotlinx.coroutines.runBlocking
 import okhttp3.HttpUrl
 
 /**
@@ -17,20 +18,10 @@ class SyncException(cause: Throwable) : Exception(cause) {
 
         // provide lambda wrappers for setting the local/remote resource
 
-        fun <T> wrapWithLocalResource(localResource: LocalResource<*>?, body: () -> T): T {
-            try {
-                return body()
-            } catch (e: SyncException) {
-                if (localResource != null)
-                    e.setLocalResourceIfNull(localResource)
-                throw e
-            } catch (e: Throwable) {
-                throw if (localResource != null)
-                    SyncException(e).setLocalResourceIfNull(localResource)
-                else
-                    e
+        fun <T> wrapWithLocalResource(localResource: LocalResource<*>?, body: () -> T): T =
+            runBlocking {
+                wrapWithLocalResourceSuspending(localResource, body)
             }
-        }
 
         suspend fun <T> wrapWithLocalResourceSuspending(localResource: LocalResource<*>?, body: suspend () -> T): T {
             try {
@@ -47,20 +38,10 @@ class SyncException(cause: Throwable) : Exception(cause) {
             }
         }
 
-        fun <T> wrapWithRemoteResource(remoteResource: HttpUrl?, body: () -> T): T {
-            try {
-                return body()
-            } catch (e: SyncException) {
-                if (remoteResource != null)
-                    e.setRemoteResourceIfNull(remoteResource)
-                throw e
-            } catch (e: Throwable) {
-                throw if (remoteResource != null)
-                    SyncException(e).setRemoteResourceIfNull(remoteResource)
-                else
-                    e
+        fun <T> wrapWithRemoteResource(remoteResource: HttpUrl?, body: () -> T): T =
+            runBlocking {
+                wrapWithRemoteResourceSuspending(remoteResource, body)
             }
-        }
 
         suspend fun <T> wrapWithRemoteResourceSuspending(remoteResource: HttpUrl?, body: suspend () -> T): T {
             try {

--- a/app/src/main/kotlin/at/bitfire/davdroid/sync/SyncManager.kt
+++ b/app/src/main/kotlin/at/bitfire/davdroid/sync/SyncManager.kt
@@ -674,7 +674,7 @@ abstract class SyncManager<ResourceType: LocalResource<*>, out CollectionType: L
      *      but not a single one (or vice versa). So only one method is more user-friendly.
      *   5. March 2020: iCloud now crashes with HTTP 500 upon CardDAV GET requests.
      */
-    protected abstract fun downloadRemote(bunch: List<HttpUrl>)
+    protected abstract suspend fun downloadRemote(bunch: List<HttpUrl>)
 
     /**
      * Locally deletes entries which are

--- a/app/src/main/kotlin/at/bitfire/davdroid/sync/SyncManager.kt
+++ b/app/src/main/kotlin/at/bitfire/davdroid/sync/SyncManager.kt
@@ -48,10 +48,10 @@ import kotlinx.coroutines.withContext
 import okhttp3.HttpUrl
 import okhttp3.RequestBody
 import java.io.IOException
-import java.io.InterruptedIOException
 import java.net.HttpURLConnection
 import java.security.cert.CertificateException
 import java.util.LinkedList
+import java.util.concurrent.CancellationException
 import java.util.concurrent.LinkedBlockingQueue
 import java.util.logging.Level
 import java.util.logging.Logger
@@ -260,9 +260,8 @@ abstract class SyncManager<ResourceType: LocalResource<*>, out CollectionType: L
                 is DeadObjectException ->
                     throw e
 
-                // sync was cancelled or account has been removed: re-throw to BaseSyncer
-                is InterruptedException,
-                is InterruptedIOException,
+                // sync was cancelled or account has been removed: re-throw to Syncer
+                is CancellationException,
                 is InvalidAccountException ->
                     throw e
 

--- a/app/src/main/kotlin/at/bitfire/davdroid/sync/SyncManager.kt
+++ b/app/src/main/kotlin/at/bitfire/davdroid/sync/SyncManager.kt
@@ -306,7 +306,7 @@ abstract class SyncManager<ResourceType: LocalResource<*>, out CollectionType: L
      *
      * @return current sync state
      */
-    protected abstract fun queryCapabilities(): SyncState?
+    protected abstract suspend fun queryCapabilities(): SyncState?
 
     /**
      * Processes locally deleted entries. This can mean:
@@ -548,7 +548,7 @@ abstract class SyncManager<ResourceType: LocalResource<*>, out CollectionType: L
      *
      * @param listRemote function to list remote resources (for instance, all since a certain sync-token)
      */
-    protected open suspend fun syncRemote(listRemote: (MultiResponseCallback) -> Unit) = coroutineScope {    // structured concurrency
+    protected open suspend fun syncRemote(listRemote: suspend (MultiResponseCallback) -> Unit) = coroutineScope {    // structured concurrency
         // download queue
         val toDownload = LinkedBlockingQueue<HttpUrl>()
         fun download(url: HttpUrl?) {
@@ -622,24 +622,27 @@ abstract class SyncManager<ResourceType: LocalResource<*>, out CollectionType: L
         download(null)
     }
 
-    protected abstract fun listAllRemote(callback: MultiResponseCallback)
+    protected abstract suspend fun listAllRemote(callback: MultiResponseCallback)
 
-    protected open fun listRemoteChanges(syncState: SyncState?, callback: MultiResponseCallback): Pair<SyncToken, Boolean> {
+    protected open suspend fun listRemoteChanges(syncState: SyncState?, callback: MultiResponseCallback): Pair<SyncToken, Boolean> {
         var furtherResults = false
 
-        val report = davCollection.reportChanges(
+        val report = runInterruptible {
+            davCollection.reportChanges(
                 syncState?.takeIf { syncState.type == SyncState.Type.SYNC_TOKEN }?.value,
                 false, null,
-                GetETag.NAME) { response, relation ->
-            when (relation) {
-                Response.HrefRelation.SELF ->
-                    furtherResults = response.status?.code == 507
+                GetETag.NAME
+            ) { response, relation ->
+                when (relation) {
+                    Response.HrefRelation.SELF ->
+                        furtherResults = response.status?.code == 507
 
-                Response.HrefRelation.MEMBER ->
-                    callback.onResponse(response, relation)
+                    Response.HrefRelation.MEMBER ->
+                        callback.onResponse(response, relation)
 
-                else ->
-                    logger.fine("Unexpected sync-collection response: $response")
+                    else ->
+                        logger.fine("Unexpected sync-collection response: $response")
+                }
             }
         }
 
@@ -702,11 +705,13 @@ abstract class SyncManager<ResourceType: LocalResource<*>, out CollectionType: L
                 SyncState(SyncState.Type.CTAG, it)
             }
 
-    private fun querySyncState(): SyncState? {
+    private suspend fun querySyncState(): SyncState? {
         var state: SyncState? = null
-        davCollection.propfind(0, GetCTag.NAME, SyncToken.NAME) { response, relation ->
-            if (relation == Response.HrefRelation.SELF)
-                state = syncState(response)
+        runInterruptible {
+            davCollection.propfind(0, GetCTag.NAME, SyncToken.NAME) { response, relation ->
+                if (relation == Response.HrefRelation.SELF)
+                    state = syncState(response)
+            }
         }
         return state
     }

--- a/app/src/main/kotlin/at/bitfire/davdroid/sync/SyncManager.kt
+++ b/app/src/main/kotlin/at/bitfire/davdroid/sync/SyncManager.kt
@@ -44,6 +44,7 @@ import dagger.hilt.android.qualifiers.ApplicationContext
 import kotlinx.coroutines.CoroutineDispatcher
 import kotlinx.coroutines.coroutineScope
 import kotlinx.coroutines.launch
+import kotlinx.coroutines.runInterruptible
 import kotlinx.coroutines.withContext
 import okhttp3.HttpUrl
 import okhttp3.RequestBody
@@ -315,14 +316,14 @@ abstract class SyncManager<ResourceType: LocalResource<*>, out CollectionType: L
      *
      * @return whether local resources have been processed so that a synchronization is always necessary
      */
-    protected open fun processLocallyDeleted(): Boolean {
+    protected open suspend fun processLocallyDeleted(): Boolean {
         var numDeleted = 0
 
         // Remove locally deleted entries from server (if they have a name, i.e. if they were uploaded before),
         // but only if they don't have changed on the server. Then finally remove them from the local address book.
         val localList = localCollection.findDeleted()
         for (local in localList) {
-            SyncException.wrapWithLocalResource(local) {
+            SyncException.wrapWithLocalResourceSuspending(local) {
                 val fileName = local.fileName
                 if (fileName != null) {
                     val lastScheduleTag = local.scheduleTag
@@ -331,13 +332,15 @@ abstract class SyncManager<ResourceType: LocalResource<*>, out CollectionType: L
 
                     val url = collection.url.newBuilder().addPathSegment(fileName).build()
                     val remote = DavResource(httpClient.okHttpClient, url)
-                    SyncException.wrapWithRemoteResource(url) {
+                    SyncException.wrapWithRemoteResourceSuspending(url) {
                         try {
-                            remote.delete(
-                                ifETag = lastETag,
-                                ifScheduleTag = lastScheduleTag,
-                                headers = pushDontNotifyHeader,
-                            ) {}
+                            runInterruptible {
+                                remote.delete(
+                                    ifETag = lastETag,
+                                    ifScheduleTag = lastScheduleTag,
+                                    headers = pushDontNotifyHeader,
+                                ) {}
+                            }
                             numDeleted++
                         } catch (_: HttpException) {
                             logger.warning("Couldn't delete $fileName from server; ignoring (may be downloaded again)")
@@ -366,7 +369,7 @@ abstract class SyncManager<ResourceType: LocalResource<*>, out CollectionType: L
         coroutineScope {    // structured concurrency
             for (local in localCollection.findDirty())
                 launch {
-                    SyncException.wrapWithLocalResource(local) {
+                    SyncException.wrapWithLocalResourceSuspending(local) {
                         uploadDirty(local)
                         numUploaded++
                     }
@@ -376,7 +379,7 @@ abstract class SyncManager<ResourceType: LocalResource<*>, out CollectionType: L
         return numUploaded > 0
     }
 
-    protected fun uploadDirty(local: ResourceType) {
+    protected suspend fun uploadDirty(local: ResourceType) {
         val existingFileName = local.fileName
 
         var newFileName: String? = null
@@ -393,14 +396,17 @@ abstract class SyncManager<ResourceType: LocalResource<*>, out CollectionType: L
 
                 val uploadUrl = collection.url.newBuilder().addPathSegment(newFileName).build()
                 val remote = DavResource(httpClient.okHttpClient, uploadUrl)
-                SyncException.wrapWithRemoteResource(uploadUrl) {
+                SyncException.wrapWithRemoteResourceSuspending(uploadUrl) {
                     logger.info("Uploading new record ${local.id} -> $newFileName")
-                    remote.put(
-                        generateUpload(local),
-                        ifNoneMatch = true,
-                        callback = readTagsFromResponse,
-                        headers = pushDontNotifyHeader
-                    )
+                    val bodyToUpload = generateUpload(local)
+                    runInterruptible {
+                        remote.put(
+                            bodyToUpload,
+                            ifNoneMatch = true,
+                            callback = readTagsFromResponse,
+                            headers = pushDontNotifyHeader
+                        )
+                    }
                 }
 
             } else /* existingFileName != null */ {     // updated resource
@@ -408,17 +414,20 @@ abstract class SyncManager<ResourceType: LocalResource<*>, out CollectionType: L
 
                 val uploadUrl = collection.url.newBuilder().addPathSegment(existingFileName).build()
                 val remote = DavResource(httpClient.okHttpClient, uploadUrl)
-                SyncException.wrapWithRemoteResource(uploadUrl) {
+                SyncException.wrapWithRemoteResourceSuspending(uploadUrl) {
                     val lastScheduleTag = local.scheduleTag
                     val lastETag = if (lastScheduleTag == null) local.eTag else null
                     logger.info("Uploading modified record ${local.id} -> $existingFileName (ETag=$lastETag, Schedule-Tag=$lastScheduleTag)")
-                    remote.put(
-                        generateUpload(local),
-                        ifETag = lastETag,
-                        ifScheduleTag = lastScheduleTag,
-                        callback = readTagsFromResponse,
-                        headers = pushDontNotifyHeader
-                    )
+                    val bodyToUpload = generateUpload(local)
+                    runInterruptible {
+                        remote.put(
+                            bodyToUpload,
+                            ifETag = lastETag,
+                            ifScheduleTag = lastScheduleTag,
+                            callback = readTagsFromResponse,
+                            headers = pushDontNotifyHeader
+                        )
+                    }
                 }
             }
         } catch (e: SyncException) {

--- a/app/src/main/kotlin/at/bitfire/davdroid/sync/TasksSyncManager.kt
+++ b/app/src/main/kotlin/at/bitfire/davdroid/sync/TasksSyncManager.kt
@@ -30,6 +30,7 @@ import dagger.assisted.Assisted
 import dagger.assisted.AssistedFactory
 import dagger.assisted.AssistedInject
 import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.runInterruptible
 import okhttp3.HttpUrl
 import okhttp3.RequestBody
 import okhttp3.RequestBody.Companion.toRequestBody
@@ -81,16 +82,18 @@ class TasksSyncManager @AssistedInject constructor(
         return true
     }
 
-    override fun queryCapabilities() =
-        SyncException.wrapWithRemoteResource(collection.url) {
+    override suspend fun queryCapabilities() =
+        SyncException.wrapWithRemoteResourceSuspending(collection.url) {
             var syncState: SyncState? = null
-            davCollection.propfind(0, MaxResourceSize.NAME, GetCTag.NAME, SyncToken.NAME) { response, relation ->
-                if (relation == Response.HrefRelation.SELF) {
-                    response[MaxResourceSize::class.java]?.maxSize?.let { maxSize ->
-                        logger.info("Calendar accepts tasks up to ${Formatter.formatFileSize(context, maxSize)}")
-                    }
+            runInterruptible {
+                davCollection.propfind(0, MaxResourceSize.NAME, GetCTag.NAME, SyncToken.NAME) { response, relation ->
+                    if (relation == Response.HrefRelation.SELF) {
+                        response[MaxResourceSize::class.java]?.maxSize?.let { maxSize ->
+                            logger.info("Calendar accepts tasks up to ${Formatter.formatFileSize(context, maxSize)}")
+                        }
 
-                    syncState = syncState(response)
+                        syncState = syncState(response)
+                    }
                 }
             }
 
@@ -110,10 +113,12 @@ class TasksSyncManager @AssistedInject constructor(
             os.toByteArray().toRequestBody(DavCalendar.MIME_ICALENDAR_UTF8)
         }
 
-    override fun listAllRemote(callback: MultiResponseCallback) {
-        SyncException.wrapWithRemoteResource(collection.url) {
+    override suspend fun listAllRemote(callback: MultiResponseCallback) {
+        SyncException.wrapWithRemoteResourceSuspending(collection.url) {
             logger.info("Querying tasks")
-            davCollection.calendarQuery("VTODO", null, null, callback)
+            runInterruptible {
+                davCollection.calendarQuery("VTODO", null, null, callback)
+            }
         }
     }
 

--- a/app/src/main/kotlin/at/bitfire/davdroid/sync/TasksSyncManager.kt
+++ b/app/src/main/kotlin/at/bitfire/davdroid/sync/TasksSyncManager.kt
@@ -122,28 +122,30 @@ class TasksSyncManager @AssistedInject constructor(
         }
     }
 
-    override fun downloadRemote(bunch: List<HttpUrl>) {
+    override suspend fun downloadRemote(bunch: List<HttpUrl>) {
         logger.info("Downloading ${bunch.size} iCalendars: $bunch")
         // multiple iCalendars, use calendar-multi-get
-        SyncException.wrapWithRemoteResource(collection.url) {
-            davCollection.multiget(bunch) { response, _ ->
-                // See CalendarSyncManager for more information about the multi-get response
-                SyncException.wrapWithRemoteResource(response.href) wrapResource@ {
-                    if (!response.isSuccess()) {
-                        logger.warning("Ignoring non-successful multi-get response for ${response.href}")
-                        return@wrapResource
+        SyncException.wrapWithRemoteResourceSuspending(collection.url) {
+            runInterruptible {
+                davCollection.multiget(bunch) { response, _ ->
+                    // See CalendarSyncManager for more information about the multi-get response
+                    SyncException.wrapWithRemoteResource(response.href) wrapResource@{
+                        if (!response.isSuccess()) {
+                            logger.warning("Ignoring non-successful multi-get response for ${response.href}")
+                            return@wrapResource
+                        }
+
+                        val iCal = response[CalendarData::class.java]?.iCalendar
+                        if (iCal == null) {
+                            logger.warning("Ignoring multi-get response without calendar-data")
+                            return@wrapResource
+                        }
+
+                        val eTag = response[GetETag::class.java]?.eTag
+                            ?: throw DavException("Received multi-get response without ETag")
+
+                        processVTodo(response.href.lastSegment, eTag, StringReader(iCal))
                     }
-
-                    val iCal = response[CalendarData::class.java]?.iCalendar
-                    if (iCal == null) {
-                        logger.warning("Ignoring multi-get response without calendar-data")
-                        return@wrapResource
-                    }
-
-                    val eTag = response[GetETag::class.java]?.eTag
-                        ?: throw DavException("Received multi-get response without ETag")
-
-                    processVTodo(response.href.lastSegment, eTag, StringReader(iCal))
                 }
             }
         }

--- a/app/src/main/kotlin/at/bitfire/davdroid/sync/worker/BaseSyncWorker.kt
+++ b/app/src/main/kotlin/at/bitfire/davdroid/sync/worker/BaseSyncWorker.kt
@@ -37,7 +37,6 @@ import at.bitfire.davdroid.ui.NotificationRegistry
 import at.bitfire.ical4android.TaskProvider
 import dagger.Lazy
 import kotlinx.coroutines.delay
-import kotlinx.coroutines.runInterruptible
 import java.util.Collections
 import java.util.logging.Level
 import java.util.logging.Logger
@@ -180,9 +179,7 @@ abstract class BaseSyncWorker(
         }
 
         // Start syncing
-        runInterruptible {
-            syncer()
-        }
+        syncer()
 
         // convert SyncResult from Syncers to worker Data
         val output = Data.Builder()

